src/Bicep.Core/Analyzers/Linter/Rules/MaxNumberAssertsRule.cs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class MaxNumberAssertsRule : LinterRuleBase
+    {
+        public new const string Code = "max-asserts";
+        public const int MaxNumber = 32;
+
+        public MaxNumberAssertsRule() : base(
+            code: Code,
+            description: CoreResources.MaxNumberAssertsRuleDescription,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}"),
+            diagnosticLevel: DiagnosticLevel.Error)
+        { }
+
+        public override string FormatMessage(params object[] values)
+        {
+            return string.Format(CoreResources.MaxNumberAssertsRuleMessageFormat, values);
+        }
+
+        override public IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model, DiagnosticLevel diagnosticLevel)
+        {
+            if (model.Root.AssertDeclarations.Count() > MaxNumber)
+            {
+                var firstItem = model.Root.AssertDeclarations.First();
+                return new IDiagnostic[] { CreateDiagnosticForSpan(diagnosticLevel, firstItem.NameSource.Span, MaxNumber) };
+            }
+            return Enumerable.Empty<IDiagnostic>();
```

Comment:
```suggestion
            if (model.Root.AssertDeclarations.Count() > MaxNumber)
            {
                var firstItem = model.Root.AssertDeclarations.First();
                yield return CreateDiagnosticForSpan(diagnosticLevel, firstItem.NameSource.Span, MaxNumber);
            }
```

[nit]: You can remove some of the boilerplate by making this [an iterator method](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/yield).